### PR TITLE
[WIP] Add annotation based commands

### PR DESCRIPTION
**Pull request by jamierocks**
[Common](https://github.com/SpongePowered/SpongeCommon/pull/656)

Example: 
```java
    @Command(
        aliases = {"actions"}
    )
    @TranslatableDescription(
        description = "some.translation.key.description",
        usage = "some.translation.key.usage",
        help = "some.translation.key.help"
    )
    @Group("kittens") // /kittens actions
    @Group("cats") // /cats actions
    @Permission({
        "some.permission",
        "some.other.permission"
    })
    public CommandResult example() {
        return CommandResult.empty();
    }

    @Command(
        aliases = {"pet"}
    )
    @PlainDescription(
        description = "Pet the kitty",
        usage = "/kittens pet",
        help = "how2helppet?"
    )
    @Group({"kittens", "actions"}) // /kittens actions pet
    @Group({"cats", "actions"}) // /cats actions pet
    @Permission({
        "some.permission",
        "some.other.permission"
    })
    public CommandResult anotherExample() {
        return CommandResult.empty();
    }
```

### CODE REVIEW

File: src/main/java/org/spongepowered/api/command/Command.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+     * @return the command's parent.
+     */
+    String parent() default "";
+
+    /**
+     * Gets the name of the command.
+     *
+     * @return the command's name.
+     */
+    String name();
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/command/Command.java]: No need for this, just require `aliases`.

File: src/main/java/org/spongepowered/api/command/Command.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface Command {
+
+    /**
+     * Gets the parent of this command.
+     *
+     * @return the command's parent.
+     */
+    String parent() default "";
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/command/Command.java]: Why?

File: src/main/java/org/spongepowered/api/command/Command.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    String usage() default "";
+
+    /**
+     * Gets all the permissions associated with this command.
+     * The {@link CommandSource} will have to have all permissions to be able to use the command.
+     * Not having any permission, means that all users can use the command, use with caution!
+     *
+     * @return the command's permissions.
+     */
+    String[] permissions() default {};
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/command/Command.java]: Why part of `@Command`?

**Comment by kashike** [comment]
I'd personally rather see something more versatile like [Intake](https://github.com/sk89q/Intake).

File: src/main/java/org/spongepowered/api/command/Command.java (hunk 2 above)
Comment (jamierocks) [inline src/main/java/org/spongepowered/api/command/Command.java]: For subcommands.

**Comment by zml2008** [comment]
See #1171 for pre-PR discussion. As mentioned there, I don't think this should be considered done without support for using CommandElements from GenericArguments/etc as method arguments, but it's a good starting point.

File: src/main/java/org/spongepowered/api/command/Command.java (hunk 3 above)
Comment (jamierocks) [inline src/main/java/org/spongepowered/api/command/Command.java]: Are you proposing a separate annotation?

Comment (zml2008) [inline src/main/java/org/spongepowered/api/command/Command.java]: I don't see why a separate annotation is necessary -- I see this wrapping around CommandSpec, which takes permission arguments in its builder.

**Comment by zml2008** [comment]
Another thing to think about is how you'd go about implementing child commands in this -- since that's useful for a lot of command structures.

**Comment by jamierocks** [comment]
Currently I use:
```java
    @Command(parent = "example",
            name = "sub")
    public CommandResult exampleSubCommand(CommandSource source, CommandArgs args) {
        source.sendMessage(Text.of("This is a sub command."));
        return CommandResult.success();
    }
```

File: src/main/java/org/spongepowered/api/command/Command.java (hunk 1 above)
Comment (jamierocks) [inline src/main/java/org/spongepowered/api/command/Command.java]: Yeah - not all too sure why I did this :p

File: src/main/java/org/spongepowered/api/command/TabComplete.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.command;
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/command/TabComplete.java]: Why isn't this in the `org.spongepowered.api.command.annotation` package?

File: src/main/java/org/spongepowered/api/command/TabComplete.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * This annotation is applied to methods, to handle tab-completion.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface TabComplete {
+
+    /**
+     * Gets the name of the command.
+     *
+     * @return the command's name.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/command/TabComplete.java]: Capital 'T' in 'The'
Don't add period to end of `@return` statements

**Comment by kashike** [comment]
This depends on #1088, so once @windy1 finishes it and it is merged we're good to go here.

**Comment by jamierocks** [comment]
see @kashike's pr

**Pull request closed without merging**